toolchain/driver/compile_subcommand.cpp

```diff
@@ -0,0 +1,757 @@
+
+  b.AddFlag(
+      {
+          .name = "dump-shared-values",
+          .help = R"""(
+Dumps shared values. These aren't owned by any particular file or phase.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_shared_values); });
+  b.AddFlag(
+      {
+          .name = "dump-tokens",
+          .help = R"""(
+Dump the tokens to stdout when lexed.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_tokens); });
+  b.AddFlag(
+      {
+          .name = "dump-parse-tree",
+          .help = R"""(
+Dump the parse tree to stdout when parsed.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_parse_tree); });
+  b.AddFlag(
+      {
+          .name = "preorder-parse-tree",
+          .help = R"""(
+When dumping the parse tree, reorder it so that it is in preorder rather than
+postorder.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&preorder_parse_tree); });
+  b.AddFlag(
+      {
+          .name = "dump-raw-sem-ir",
+          .help = R"""(
+Dump the raw JSON structure of SemIR to stdout when built.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_raw_sem_ir); });
+  b.AddFlag(
+      {
+          .name = "dump-sem-ir",
+          .help = R"""(
+Dump the SemIR to stdout when built.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_sem_ir); });
+  b.AddFlag(
+      {
+          .name = "builtin-sem-ir",
+          .help = R"""(
+Include the SemIR for builtins when dumping it.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&builtin_sem_ir); });
+  b.AddFlag(
+      {
+          .name = "dump-llvm-ir",
+          .help = R"""(
+Dump the LLVM IR to stdout after lowering.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_llvm_ir); });
+  b.AddFlag(
+      {
+          .name = "dump-asm",
+          .help = R"""(
+Dump the generated assembly to stdout after codegen.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_asm); });
+  b.AddFlag(
+      {
+          .name = "dump-mem-usage",
+          .help = R"""(
+Dumps the amount of memory used.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&dump_mem_usage); });
+  b.AddFlag(
+      {
+          .name = "prelude-import",
+          .help = R"""(
+Whether to use the implicit prelude import. Enabled by default.
+)""",
+      },
+      [&](auto& arg_b) {
+        arg_b.Default(true);
+        arg_b.Set(&prelude_import);
+      });
+  b.AddStringOption(
+      {
+          .name = "exclude-dump-file-prefix",
+          .value_name = "PREFIX",
+          .help = R"""(
+Excludes files with the given prefix from dumps.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Set(&exclude_dump_file_prefix); });
+  b.AddFlag(
+      {
+          .name = "debug-info",
+          .help = R"""(
+Emit DWARF debug information.
+)""",
+      },
+      [&](auto& arg_b) {
+        arg_b.Default(true);
+        arg_b.Set(&include_debug_info);
+      });
+}
+
+auto Driver::ValidateCompileOptions(const CompileOptions& options) const
```

Comment:
I'm worried it could be quite confusing that some of the members of `Driver` aren't declared in `driver.cpp`. Even though this is a temporary measure, I think it would be helpful to have comments in `driver.h`explaining where to find the corresponding implementations.

Reply:
Done.

Reply:
That's not quite what I had in mind; I've added some suggestions.

Reply:
To be sure, #4300 (already pending) is going to remove these TODOs. Do you still want them?

Reply:
Yes, I'd still prefer to have them. You never know what commit someone's going to start their dev branch at, so all else being equal it's better to avoid committing code in a potentially confusing state, even temporarily.

Reply:
Done. To be clear, I strongly disagree, but I'll just spend the time doing  this.